config/src/main/java/org/springframework/security/config/http/OAuth2ResourceServerBeanDefinitionParser.java

```diff
@@ -0,0 +1,336 @@
+		if (!StringUtils.isEmpty(authenticationManagerResolverRef)) {
+			return new RuntimeBeanReference(authenticationManagerResolverRef);
+		}
+		BeanDefinitionBuilder authenticationManagerResolver = BeanDefinitionBuilder
+				.rootBeanDefinition(StaticAuthenticationManagerResolver.class);
+		authenticationManagerResolver.addConstructorArgValue(this.authenticationManager);
+		return authenticationManagerResolver.getBeanDefinition();
+	}
+
+	BeanMetadataElement getBearerTokenResolver(Element element) {
+		String bearerTokenResolverRef = element.getAttribute(BEARER_TOKEN_RESOLVER_REF);
+		if (StringUtils.isEmpty(bearerTokenResolverRef)) {
+			return new RootBeanDefinition(DefaultBearerTokenResolver.class);
+		} else {
+			return new RuntimeBeanReference(bearerTokenResolverRef);
+		}
+	}
+}
+
+final class JwtBeanDefinitionParser implements BeanDefinitionParser {
+	static final String DECODER_REF = "decoder-ref";
+	static final String JWK_SET_URI = "jwk-set-uri";
+	static final String JWT_AUTHENTICATION_CONVERTER_REF = "jwt-authentication-converter-ref";
+	static final String JWT_AUTHENTICATION_CONVERTER = "jwtAuthenticationConverter";
+
+	@Override
+	public BeanDefinition parse(Element element, ParserContext pc) {
+		validateConfiguration(element, pc);
+
+		BeanDefinitionBuilder jwtProviderBuilder =
+				BeanDefinitionBuilder.rootBeanDefinition(JwtAuthenticationProvider.class);
+		jwtProviderBuilder.addConstructorArgValue(getDecoder(element));
+		jwtProviderBuilder.addPropertyValue(JWT_AUTHENTICATION_CONVERTER, getJwtAuthenticationConverter(element));
+
+		return jwtProviderBuilder.getBeanDefinition();
+	}
+
+	void validateConfiguration(Element element, ParserContext pc) {
+		boolean usesDecoder = element.hasAttribute(DECODER_REF);
+		boolean usesJwkSetUri = element.hasAttribute(JWK_SET_URI);
+
+		if (usesDecoder == usesJwkSetUri) {
+			pc.getReaderContext().error
+					("Please specify either decoder-ref or jwk-set-uri.", element);
+		}
+	}
+
+	Object getDecoder(Element element) {
+		String decoderRef = element.getAttribute(DECODER_REF);
+		if (!StringUtils.isEmpty(decoderRef)) {
+			return new RuntimeBeanReference(decoderRef);
+		}
+
+		BeanDefinitionBuilder builder = BeanDefinitionBuilder
+				.rootBeanDefinition(NimbusJwtDecoderJwkSetUriFactoryBean.class);
+		builder.addConstructorArgValue(element.getAttribute(JWK_SET_URI));
+		return builder.getBeanDefinition();
+	}
+
+	Object getJwtAuthenticationConverter(Element element) {
+		String jwtDecoderRef = element.getAttribute(JWT_AUTHENTICATION_CONVERTER_REF);
+		if (!StringUtils.isEmpty(jwtDecoderRef)) {
+			return new RuntimeBeanReference(jwtDecoderRef);
+		}
+
+		return new JwtAuthenticationConverter();
+	}
+
+	JwtBeanDefinitionParser() {}
+}
+
+final class OpaqueTokenBeanDefinitionParser implements BeanDefinitionParser {
+	static final String INTROSPECTOR_REF = "introspector-ref";
+	static final String INTROSPECTION_URI = "introspection-uri";
+	static final String CLIENT_ID = "client-id";
```

Comment:
I wonder if this should be renamed to `introspection-client-id` - similar for `client-secret`?

Reply:
These match the names of the properties in Spring Boot, which is why I went with these names. Also, the repetition of the name feels just a little bit cumbersome when typing.

I suppose I'm open, though - can you elaborate on why you feel this is important?

Reply:
I missed the fact that these properties are same as in Spring Boot. Let's keep as-is then.